RecoEgamma/EgammaTools/src/HGCalIsoProducer.cc

```diff
@@ -0,0 +1,115 @@
+/*
+ * HGCalIsoProducer.cc
+ *
+ *  Created on: 13 Oct 2017
+ *      Author: jkiesele
+ */
+
+#include "DataFormats/Math/interface/deltaR.h"
+#include <stdexcept>
+#include "../interface/HGCalIsoProducer.h"
+
+HGCalIsoProducer::HGCalIsoProducer():dr2_(0.15*0.15),mindr2_(0),rechittools_(nullptr),debug_(false),nlayers_(30){
+    setNRings(5);
+}
+
+HGCalIsoProducer::~HGCalIsoProducer(){
+}
+
+void HGCalIsoProducer::setRecHits(
+        edm::Handle<HGCRecHitCollection> hitsEE,
+        edm::Handle<HGCRecHitCollection> hitsFH,
+        edm::Handle<HGCRecHitCollection> hitsBH
+     ){
+    recHitsEE_=hitsEE;
+    recHitsFH_=hitsFH;
+    recHitsBH_=hitsBH;
+
+    if(!rechittools_)
+        throw std::runtime_error("HGCalIsoProducer::produceHGCalIso: rechittools not set");
+    
+    hitEtaPhiCache_.clear();
+    hitEtaPhiCache_.reserve(recHitsEE_->size()+recHitsFH_->size()+recHitsBH_->size());
+
+    // Since HGCal is not projective and the rechits don't cache any
+    // eta,phi, we make our own here
+    auto makeEtaPhiPair = [this](const auto& hit) {
+        const GlobalPoint position = rechittools_->getPosition(hit.id());
+        float eta=rechittools_->getEta(position, 0); //assume vertex at z=0
+        float phi=rechittools_->getPhi(position);
+        return std::make_pair(eta, phi);
+    };
+
+    for(const auto& hit : *recHitsEE_) hitEtaPhiCache_.push_back(makeEtaPhiPair(hit));
+    for(const auto& hit : *recHitsFH_) hitEtaPhiCache_.push_back(makeEtaPhiPair(hit));
+    for(const auto& hit : *recHitsBH_) hitEtaPhiCache_.push_back(makeEtaPhiPair(hit));
+}
+
+void HGCalIsoProducer::produceHGCalIso(const reco::CaloClusterPtr & seed){
+
+    if(!rechittools_)
+        throw std::runtime_error("HGCalIsoProducer::produceHGCalIso: rechittools not set");
+
+    for(auto& r:isoringdeposits_)
+        r=0;
+
+    // compiler couldn't figure out seed is const and
+    // recomputed log(atan2()) every time...
```

Comment:
are these comments necessary?

Reply:
It explains why I didn't just capture `seed` in the lambda.

Reply:
got it.
please make it more explicit or drop the comment. (because it's a comment about something that does not exist)